core/src/main/java/org/apache/calcite/rex/RexProgram.java

```diff
@@ -821,6 +824,29 @@ public RexProgram normalize(RexBuilder rexBuilder, boolean simplify) {
         : null);
   }
 
+  /**
+   * Returns a partial mapping of a set of project expressions.
+   *
+   * <p>The mapping is an inverse function.
+   * Every target has a source field, but
+   * a source might have 0, 1 or more targets.
+   * Project expressions that do not consist of
+   * a mapping are ignored.
+   *
+   * @param inputFieldCount Number of input fields
+   * @return Mapping of a set of project expressions, never null
+   */
+  public Mappings.TargetMapping getPartialMapping(int inputFieldCount) {
+    Mappings.TargetMapping mapping =
+        Mappings.create(MappingType.INVERSE_FUNCTION,
+            inputFieldCount, projects.size());
+    for (Ord<RexLocalRef> exp : Ord.zip(projects)) {
+      RexInputRef inputRef = (RexInputRef) exprs.get(exp.e.index);
```

Comment:
Shall we call `RexProgram#expandLocalRef` for the expansion ?

Reply:
Yes, that's better.
Thanks for remindering me.